### ISSUE TITLE
A freezed class with union types that all include another union type should have that union type base type available without casting

### ISSUE THREAD

**Issue by hacker1024** [rrousselGit/freezed#425]
**Is your feature request related to a problem? Please describe.**
Consider the following:

```dart
class GeneralMediaDetails {/* ... */}

@freezed
class Media with _$Media {
  const factory Media.station(/* ... */) = StationMedia;
  const factory Media.track(/* ... */) = TrackMedia;
}

@freezed
class MediaDetailSet with _$MediaDetailSet {
  const factory MediaDetailSet.station(GeneralMediaDetails generalDetails, StationMedia media) = StationMediaDetailSet;
  const factory MediaDetailSet.track(GeneralMediaDetails generalDetails, TrackMedia media) = TrackMediaDetailSet;
}
```

The following would work just fine:

```dart
void printGeneralDetails(MediaDetailSet detailSet) {
  print(detailSet.generalDetails);
}
```

But this would not:

```dart
void printMedia(MediaDetailSet detailSet) {
  print(detailSet.media);
}
```

`media` isn't part of the base `MediaDetailSet` type.
It can be, though - both unions have a `Media media` property.

**Describe the solution you'd like**
Freezed should make downcasted union types available in this scenario.

**Describe alternatives you've considered**
Casting must be done to access the field.

**Additional context**

**Comment by rrousselGit** [comment]
That is already the case.

**Comment by hacker1024** [comment]
It isn't - copy that code and see for yourself.

**Comment by rrousselGit** [comment]
It is. 

That has been part of the readme for a long time
https://github.com/rrousselGit/freezed#shared-properties

**Comment by hacker1024** [comment]
I have read the readme. Even though the `media` property is defined on all constructors, it is not accessible without casting. Perhaps the common base type is not being recognised?

As I said, the code in my first post does not generate a `media` getter in `_$MediaDetailSet`.

**Comment by rrousselGit** [comment]
Oh wait, I think I know what the problem is.

What's `GeneralMediaDetails` in your snippet? My guess is that it is similar to `StationMedia`?

**Comment by rrousselGit** [comment]
I've just tested your code and I can confirm that this behaves as you described:

![Screenshot 2021-04-13 at 17 16 19](https://user-images.githubusercontent.com/20165741/114585845-f88f8d80-9c7b-11eb-80ff-2f386f463edf.png)

**Comment by harkairt** [comment]
I think OP is talking about `media` not being there with `Media` type, although both union types have a `media` named prop, but they have it with a different subclass type. 

The difference between `media` and `generalDetails` (which is there in your screenshot) is that both types have `generalDetails` with type `GeneralMediaDetails`, but the `media` is just sharing a common baseclass `Media` while the actual class differs. 

I've once wanted to have this functionality myself as well

**Comment by rrousselGit** [comment]
Oh I see, I got confused.

This is difficult. `StationMedia` & `TrackMedia` do not exist yet when `MediaDetailSet` is parsed

**Comment by rrousselGit** [comment]
I'll close this as I don't plan to work on it.

Fixing it with the way things are currently working would be very tedious – and this is a fairly rare use-case IMO.  
At the same time, metaprogramming will fix the problem without me having to do anything.

**Comment by hacker1024** [comment]
For anyone looking at this now, this appears to have been implemented in v2.3.0 with https://github.com/rrousselGit/freezed/commit/e0e09ac728e66b44f23cd8109a58c77dee9a7667.
